Query updates

charlottekostelic opened the pull request:
Edited to reflect changes:

Added option for users to add automatic retries to requests
- added `total_retries`, `backoff_factor`, `status_forcelist`, and `allowed_methods` as args in `WorldcatSession` and `MetadataSession`
- default number of retries is 0
- added tests to all modules and live tests of retries to `test_metadata_api.py`

Somewhat related: there's an open enhancement request for `requests` asking to be able to refresh authentication header on retry. That functionality doesn't currently exist and it doesn't look like it's being worked on
https://github.com/psf/requests/issues/5975

klinga commented on bookops_worldcat/query.py:
```diff
@@ -49,12 +50,16 @@ def __init__(
         if not isinstance(prepared_request, PreparedRequest):
             raise TypeError("Invalid type for argument 'prepared_request'.")
 
+        # allow session to retry a request up to 3 times
```
I think a better place to set up retires is `_session.WorldcatSession` class. The default number of attempts can be as given, but it would be a good strategy to give control of this element to users, i.e. instating a session should allow to pass max_retries arg. Why `WorldcatSession`? This way it will be required only once to mount the `Retry` on the session, instead of doing it in each request. If expanded to Search API these rules will apply to these sessions as well. The other arguments (`backoff_factor` & `status_forelist`) are probably OK to be constant. 


-

klinga reviewed (commented):
@charlottekostelic in the description you mention a removal of a test for a stale token from `test_query` but I don't see such changes in this PR. We should be testing for a stale token in this module since the check if it is still valid happens only here. Stale token tests were somewhat redundant in the `test_metadata_api` and it's OK to skip them there.

Also (this may be complicated but useful), it would be smart to have a test in the `test_query` to test for retries (if they are being activated appropriately, especially on timeouts since these are common with the service).

charlottekostelic commented:
> @charlottekostelic in the description you mention a removal of a test for a stale token from `test_query` but I don't see such changes in this PR. We should be testing for a stale token in this module since the check if it is still valid happens only here. Stale token tests were somewhat redundant in the `test_metadata_api` and it's OK to skip them there.

I removed the stale token test and that lowered the test coverage (the `if session.authorization.is_expired()` block) so I added it back in but forgot to update the description.

charlottekostelic commented:
@klinga I made a few changes:

- Changed where retries are mounted so they are now in `WorldcatSession` rather than `Query`
- Added `RetryError` as an exception for `Query`
- Added additional tests

klinga commented on bookops_worldcat/_session.py:
```diff
@@ -35,6 +36,12 @@ def __init__(
         super().__init__()
 
+        # allow session to retry a request up to 3 times
+        retries = Retry(
+            total=3, backoff_factor=0.5, status_forcelist=[406, 429, 500, 502, 503, 504]
+        )
+        self.mount("https://", requests.adapters.HTTPAdapter(max_retries=retries))
```
@charlottekostelic please take a look at `requests.adapters.HTTPAdapter` code [here](https://github.com/psf/requests/blob/eea3bbf9ac635f465ee6c9903dc57c677952dafd/src/requests/adapters.py#L135C1-L145C59). If I'm reading it correctly `requests` dumbs down what is available in urllib3's `Retry` and only one argument is considered: `max_retries`. `backoff_factor` & `status_foreclist` will be ignored.
See [here](https://github.com/psf/requests/blob/eea3bbf9ac635f465ee6c9903dc57c677952dafd/src/requests/adapters.py#L121-L124) for an example how to mount the adapter and retries.

klinga replied:
Since it appears there is no control what HTTP error codes trigger a retry, we should avoid blindly setting any retries. Let's leave it to users: `WorldcatSession` should take `max_retries` argument when instantiating with a default value 0.

klinga replied:
```python
class WorldcatSession(requests.Session):
    def __init__(self, authorization, agent, timeout, max_retries):
        ....
```
It would be nice to put retries logic below timeouts and just before `self._update_authorization()`.
The argument `max_retries` should be then added to `MetadataSession`, and passed to `WorldcatSession`. Something like this:
```python
super().__init__(authorization, agent=agent, timeout=timeout, max_retries=max_retries`
```
`max_retries` in `MetadataSession` should also have default value 0.

charlottekostelic replied:
> @charlottekostelic apologies for not catching this earlier. Please take a look at `requests.adapters.HTTPAdapter` code [here](https://github.com/psf/requests/blob/eea3bbf9ac635f465ee6c9903dc57c677952dafd/src/requests/adapters.py#L135C1-L145C59). If I'm reading it correctly `requests` dumbs down what is available in urllib3's `Retry` and only one argument is considered: `max_retries`. `backoff_factor` & `status_forcelist` will be ignored. See [here](https://github.com/psf/requests/blob/eea3bbf9ac635f465ee6c9903dc57c677952dafd/src/requests/adapters.py#L121-L124) for an example how to mount the adapter and retries.

The Requests documentation has an example that passes additional arguments using a `Retry` object from `urllib3` [here](https://requests.readthedocs.io/en/latest/user/advanced/#example-automatic-retries). My understanding of the documentation is that Requests only accepts one argument (`max_retries`) out of the box but with it is possible to customize retry behavior by passing `max_retries` a `Retry` object. 

I just added a couple of tests to `test_session.py` that I think confirm this.

I like the idea of letting users decide whether to use automatic retries. It could have a default of 0 retries but allow users to customize the number of retries and behavior.

klinga replied:
Ha! Can you take a look at the `requests` code and figure out if that really is implemented? I suspect I missed things when `.mount()` is called?

charlottekostelic replied:
The [doc string](https://github.com/psf/requests/blob/eea3bbf9ac635f465ee6c9903dc57c677952dafd/src/requests/adapters.py#L110-L116) of the `max_retries` parameter says to import `Retry` from `urllib3` and pass that if you need more granular control over retry behavior. 

In the `requests` [code](https://github.com/psf/requests/blob/eea3bbf9ac635f465ee6c9903dc57c677952dafd/src/requests/adapters.py#L135C1-L145C59), `max_retries` is a `Retry` object created using the [`from_int()`](https://github.com/urllib3/urllib3/blob/main/src/urllib3/util/retry.py#L266-L284) class method. In `urllib3`, `from_int()` is used to support backwards compatibility of older retry formats.

klinga replied:
Thanks! Looks like everything was cleared up and your approach should stay.

klinga commented on bookops_worldcat/query.py:
```diff
@@ -53,8 +52,6 @@ def __init__(
         if session.authorization.is_expired():
             session._get_new_access_token()
 
-        self.response = None
```
Not sure what I was thinking to have this line. Does not make sense now.
Thanks for catching it.

klinga commented on tests/test_session.py:
```diff
@@ -38,3 +39,7 @@ def test_default_timeout(self, mock_token):
+
+    def test_adapter(self, mock_token):
+        with WorldcatSession(mock_token) as session:
+            assert isinstance(session.adapters["https://"], HTTPAdapter)
```
I don't think this assertion will work - if you look at `requests.Session` [here](https://github.com/psf/requests/blob/eea3bbf9ac635f465ee6c9903dc57c677952dafd/src/requests/sessions.py#L448) you will see that it has two adapters mounted by default (http & https). Your code will mount a custom adapter with retries instead of the default one (I think, haven't tested it). Above assertion will always be True.
Try something like that instead:
```python
    def test_no_retry_adapter(self, mock_token):
        with WorldcatSession(mock_token) as session:
            assert session.adapters["https://"].max_retries.total == 0

    def test_retry_adapter(self, mock_token):
        with WorldcatSession(mock_token, max_retries=3) as session:
            assert session.adapters["https://"].max_retries.total == 3
```

klinga commented on tests/test_query.py:
```diff
@@ -143,10 +141,32 @@ def test_query_connection_exception(stub_session, mock_connection_error):
     assert "Unexpected request error: <class 'Exception'>" in str(exc.value)
+
+
+def test_query_timeout_retry(stub_session, caplog):
+    req = Request("GET", "https://foo.org")
+    prepped = stub_session.prepare_request(req)
+    with pytest.raises(WorldcatRequestError):
+        Query(stub_session, prepped)
+
+    assert "Retry(total=0, " in caplog.records[2].message
+    assert "Retry(total=1, " in caplog.records[1].message
+    assert "Retry(total=2, " in caplog.records[0].message
```
If the default `max_retry` is 0 this will fail. It would be good to test two scenarios: default - no retries (and assertion testing a retry was NOT attempted (`assert "Retry(" not in caplog`) and with set `max_retries` like in your example. You will not be able to use `stub_session` for the latter. Instead instantiate `MetadataSession` with `mock_token` and pass `max_retries` arg.

klinga reviewed (changes requested):
@charlottekostelic further changes are needed in `_session`, `metadata_api`  and their tests.
Please see my comments for the details. Let me know if my explanations weren't clear.

klinga commented on bookops_worldcat/_session.py:
```diff
@@ -24,6 +25,10 @@ def __init__(
         ),
+        total_retries: int = 0,
+        backoff_factor: float = 0,
+        status_forcelist: Optional[List[int]] = [],
```
Does raises mypy errors? [Optional docs](https://docs.python.org/3.8/library/typing.html#typing.Optional).
My understanding of `Optional` is that it can be some Python type or None, equivalent to Union[List[int], None]

klinga replied:
Basically I'm asking if the default value in line 30 should be `None` instead of `[]`. If the latter, `Optional` can be dropped.

charlottekostelic replied:
For some reason this doesn't raise any mypy errors but it seems like it should. It also isn't working how I thought it would if the session is passed `total_retries` without `status_forcelist`. 

I'll fix this in my next PR.

klinga replied:
thanks for looking into it!

klinga commented on bookops_worldcat/metadata_api.py:
```diff
@@ -24,6 +24,10 @@ def __init__(
         timeout: Union[int, float, Tuple[int, int], Tuple[float, float], None] = None,
+        total_retries: int = 0,
+        backoff_factor: float = 0,
+        status_forcelist: Optional[List[int]] = [],
```
Does mypy agree?

klinga commented on tests/test_query.py:
```diff
@@ -143,10 +141,41 @@ def test_query_connection_exception(stub_session, mock_connection_error):
     assert "Unexpected request error: <class 'Exception'>" in str(exc.value)
+
+
+def test_query_timeout_retry(stub_retry_session, caplog):
+    req = Request("GET", "https://foo.org")
+    prepped = stub_retry_session.prepare_request(req)
+    with pytest.raises(WorldcatRequestError):
+        Query(stub_retry_session, prepped)
+
+    assert "Retry(total=0, " in caplog.records[2].message
+    assert "Retry(total=1, " in caplog.records[1].message
+    assert "Retry(total=2, " in caplog.records[0].message
+
+
+def test_query_timeout_no_retry(stub_session, caplog):
+    req = Request("GET", "https://foo.org")
+    prepped = stub_session.prepare_request(req)
+    with pytest.raises(WorldcatRequestError):
+        Query(stub_session, prepped)
+
+    assert "Retry" not in caplog.records
```
👏 Awesome!

klinga commented on tests/test_session.py:
```diff
@@ -38,3 +38,22 @@ def test_default_timeout(self, mock_token):
     def test_custom_timeout(self, mock_token):
         with WorldcatSession(mock_token, timeout=1) as session:
             assert session.timeout == 1
+
+    def test_default_adapter(self, mock_token):
+        with WorldcatSession(mock_token) as session:
+            assert session.adapters["https://"].max_retries.total == 0
+
+    def test_adapter_retries(self, mock_token):
+        with WorldcatSession(
+            authorization=mock_token,
+            total_retries=3,
+            backoff_factor=0.5,
+            status_forcelist=[500, 502, 503, 504],
+            allowed_methods=["GET", "POST", "PUT"],
+        ) as session:
+            assert session.adapters["https://"].max_retries.status_forcelist == [
+                500,
+                502,
+                503,
+                504,
+            ]
```
Looks great!

klinga reviewed (approved):
Apart from two instances of typing question I have this looks good.
